beacon-chain/rpc/prysm/v1alpha1/validator/proposer_attestations.go

```diff
@@ -309,3 +317,211 @@ func (vs *Server) deleteAttsInPool(ctx context.Context, atts []ethpb.Att) error
+	previousEpoch := slots.ToEpoch(currentSlot)
+	if previousEpoch >= 1 {
+		previousEpoch = previousEpoch.Sub(1)
+	}
+	attestationSlot := att.GetData().Slot
+	attestationEpoch := slots.ToEpoch(attestationSlot)
+	return attestationEpoch == previousEpoch
+}
+
+// filterCurrentEpochAttestationByForkchoice filters attestations from the current epoch based on fork choice conditions.
+// Returns true if all of the following conditions are met:
+// 1. The attestation beacon block root is for a slot in the previous epoch (according to fork choice).
+// 2. The attestation target root is the same as the attestation beacon block root.
+// 3. The common ancestor of the head block and the attestation beacon block root is from the previous epoch.
+func (vs *Server) filterCurrentEpochAttestationByForkchoice(ctx context.Context, att ethpb.Att, headRoot [32]byte) (bool, error) {
+	if !vs.isAttestationFromCurrentEpoch(att) {
+		return false, nil
+	}
+
+	attTargetRoot := [32]byte(att.GetData().Target.Root)
+	attBlockRoot := [32]byte(att.GetData().BeaconBlockRoot)
+	if attBlockRoot != attTargetRoot {
+		return false, nil
+	}
+
+	slot, err := vs.ForkchoiceFetcher.RecentBlockSlot(attBlockRoot)
+	if err != nil {
+		return false, err
+	}
+	epoch := slots.ToEpoch(slot)
+	prevEpoch := slots.ToEpoch(vs.TimeFetcher.CurrentSlot())
+	if prevEpoch >= 1 {
+		prevEpoch = prevEpoch.Sub(1)
+	}
+	if epoch != prevEpoch {
+		return false, nil
+	}
+
+	_, slot, err = vs.ForkchoiceFetcher.CommonAncestor(ctx, headRoot, attBlockRoot)
+	if err != nil {
+		return false, err
+	}
+	epoch = slots.ToEpoch(slot)
+	return epoch == prevEpoch, nil
+}
+
+// filterCurrentEpochAttestationByTarget returns true if an attestation from the current epoch matches the fork choice target view.
+// The conditions checked are:
+// 1. The attestation's target epoch matches the forkchoice target epoch.
+// 2. The attestation's target root matches the forkchoice target root.
+func (vs *Server) filterCurrentEpochAttestationByTarget(att ethpb.Att, targetRoot [32]byte, targetEpoch primitives.Epoch) (bool, error) {
+	if !vs.isAttestationFromCurrentEpoch(att) {
+		return false, nil
+	}
+
+	attTargetRoot := [32]byte(att.GetData().Target.Root)
+	return att.GetData().Target.Epoch == targetEpoch && attTargetRoot == targetRoot, nil
+}
+
+// filterPreviousEpochAttestationByTarget returns true if an attestation from the previous epoch matches the fork choice previous target view.
+// The conditions checked are:
+// 1. The attestation's target epoch matches the forkchoice previous target epoch.
+// 2. The attestation's target root matches the forkchoice previous target root.
+func (vs *Server) filterPreviousEpochAttestationByTarget(att ethpb.Att, targetRoot [32]byte, targetEpoch primitives.Epoch) (bool, error) {
+	if !vs.isAttestationFromPreviousEpoch(att) {
+		return false, nil
+	}
+
+	attTargetRoot := [32]byte(att.GetData().Target.Root)
+	return att.GetData().Target.Epoch == targetEpoch && attTargetRoot == targetRoot, nil
+}
```

Comment:
Why not pass a `*checkpoint` to these methods since anyway you are comparing explicitly checkpoints?